### PR TITLE
[EWPP-1539] EWPP-1661: Adds Event Programme referencable corporate entity for Event CT. 

### PR DESCRIPTION
**Pull request by 22Alexandra**

### CODE REVIEW

File: modules/oe_content_event/config/install/field.storage.node.oe_event_programme.yml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+id: node.oe_event_programme
+field_name: oe_event_programme
+entity_type: node
+type: entity_reference_revisions
+settings:
+  target_type: event_programme
+module: entity_reference_revisions
+locked: false
+cardinality: -1
+translatable: true
```
Comment (yenyasinn) [inline modules/oe_content_event/config/install/field.storage.node.oe_event_programme.yml]:
```suggestion
translatable: false
```

File: modules/oe_content_event/config/post_updates/20001_event_v2_fields/field.storage.node.oe_event_programme.yml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+id: node.oe_event_programme
+field_name: oe_event_programme
+entity_type: node
+type: entity_reference_revisions
+settings:
+  target_type: event_programme
+module: entity_reference_revisions
+locked: false
+cardinality: -1
+translatable: true
```
Comment (yenyasinn) [inline modules/oe_content_event/config/post_updates/20001_event_v2_fields/field.storage.node.oe_event_programme.yml]:
```suggestion
translatable: false
```

File: modules/oe_content_event/modules/oe_content_event_event_programme/oe_content_event_event_programme.module [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+ */
+
+declare(strict_types = 1);
+
+/**
+ * Implements hook_theme().
+ */
+function oe_content_event_event_programme_theme(): array {
+  return [
+    'event_programme' => [
```
Comment (yenyasinn) [inline modules/oe_content_event/modules/oe_content_event_event_programme/oe_content_event_event_programme.module]:
```suggestion
    'oe_event_programme' => [
```

File: modules/oe_content_event/modules/oe_content_event_event_programme/oe_content_event_event_programme.module [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+      'render element' => 'elements',
+    ],
+  ];
+}
+
+/**
+ * Implements hook_preprocess_HOOK().
+ */
+function template_preprocess_event_programme(array &$variables) {
+  oe_content_entity_default_preprocess($variables, 'event_programme');
```
Comment (yenyasinn) [inline modules/oe_content_event/modules/oe_content_event_event_programme/oe_content_event_event_programme.module]:
```suggestion
  oe_content_entity_default_preprocess($variables, 'oe_event_programme');
```

File: modules/oe_content_event/modules/oe_content_event_event_programme/oe_content_event_event_programme.module [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+/**
+ * Implements hook_theme_suggestions_HOOK().
+ */
+function oe_content_event_event_programme_theme_suggestions_event_programme(array $variables): array {
+  return oe_content_entity_default_theme_suggestion($variables, 'event_programme');
```
Comment (yenyasinn) [inline modules/oe_content_event/modules/oe_content_event_event_programme/oe_content_event_event_programme.module]:
```suggestion
  return oe_content_entity_default_theme_suggestion($variables, 'oe_event_programme');
```

File: modules/oe_content_event/modules/oe_content_event_event_programme/src/Entity/ProgrammeItem.php [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+use Drupal\oe_content_entity\Entity\CorporateEntityBase;
+
+/**
+ * Defines the Programme Item entity.
+ *
+ * @ingroup oe_content_event_event_programme
+ *
+ * @ContentEntityType(
+ *   id = "event_programme",
```
Comment (yenyasinn) [inline modules/oe_content_event/modules/oe_content_event_event_programme/src/Entity/ProgrammeItem.php]:
```suggestion
 *   id = "oe_event_programme",
```
Even @ademarco defined machine name as `event_programme`, we have to prefix this library entities with prefix `oe_`

File: modules/oe_content_event/modules/oe_content_event_event_programme/src/Entity/ProgrammeItem.php [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+ *       "default" = "Drupal\oe_content_entity\Form\CorporateEntityForm",
+ *       "add" = "Drupal\oe_content_entity\Form\CorporateEntityForm",
+ *       "edit" = "Drupal\oe_content_entity\Form\CorporateEntityForm",
+ *       "delete" = "Drupal\Core\Entity\ContentEntityDeleteForm",
+ *     },
+ *     "route_provider" = {
+ *       "html" = "Drupal\oe_content_entity\Routing\CorporateEntityRouteProvider",
+ *     },
+ *   },
+ *   base_table = "event_programme",
```
Comment (yenyasinn) [inline modules/oe_content_event/modules/oe_content_event_event_programme/src/Entity/ProgrammeItem.php]:
```suggestion
 *   base_table = "oe_event_programme",
```

File: modules/oe_content_event/modules/oe_content_event_event_programme/src/Entity/ProgrammeItem.php [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+ *   data_table = "event_programme_field_data",
```
Comment (yenyasinn) [inline modules/oe_content_event/modules/oe_content_event_event_programme/src/Entity/ProgrammeItem.php]:
```suggestion
 *   data_table = "oe_event_programme_field_data",
```

File: modules/oe_content_event/modules/oe_content_event_event_programme/src/Entity/ProgrammeItem.php [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+ *   revision_table = "event_programme_revision",
```
Comment (yenyasinn) [inline modules/oe_content_event/modules/oe_content_event_event_programme/src/Entity/ProgrammeItem.php]:
```suggestion
 *   revision_table = "oe_event_programme_revision",
```

File: modules/oe_content_event/modules/oe_content_event_event_programme/src/Entity/ProgrammeItem.php [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+ *   revision_data_table = "event_programme_field_revision",
```
Comment (yenyasinn) [inline modules/oe_content_event/modules/oe_content_event_event_programme/src/Entity/ProgrammeItem.php]:
```suggestion
 *   revision_data_table = "oe_event_programme_field_revision",
```

File: modules/oe_content_event/modules/oe_content_event_event_programme/src/Entity/ProgrammeItem.php [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+ *   },
+ *   revision_metadata_keys = {
+ *     "revision_user" = "revision_uid",
+ *     "revision_created" = "revision_timestamp",
+ *     "revision_log_message" = "revision_log"
+ *   },
+ *   bundle_entity_type = "event_programme_type",
+ *   field_ui_base_route = "entity.event_programme_type.edit_form",
+ *   links = {
+ *     "canonical" = "/admin/content/event_programme/{event_programme}",
```
Comment (yenyasinn) [inline modules/oe_content_event/modules/oe_content_event_event_programme/src/Entity/ProgrammeItem.php]:
```suggestion
 *     "canonical" = "/admin/content/oe_event_programme/{event_programme}",
```
and also all paths below

File: modules/oe_content_event/modules/oe_content_event_event_programme/src/Entity/ProgrammeItemType.php [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+namespace Drupal\oe_content_event_event_programme\Entity;
+
+use Drupal\oe_content_entity\Entity\CorporateEntityTypeBase;
+
+/**
+ * Defines the event programme type entity.
+ *
+ * @ConfigEntityType(
+ *   id = "event_programme_type",
```
Comment (yenyasinn) [inline modules/oe_content_event/modules/oe_content_event_event_programme/src/Entity/ProgrammeItemType.php]:
```suggestion
 *   id = "oe_event_programme_type",
```

File: modules/oe_content_event/modules/oe_content_event_event_programme/src/Entity/ProgrammeItemType.php [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+ *   label = @Translation("Programme Item type"),
+ *   bundle_of = "event_programme",
+ *   entity_keys = {
+ *     "id" = "id",
+ *     "label" = "label",
+ *   },
+ *   config_prefix = "event_programme_type",
```
Comment (yenyasinn) [inline modules/oe_content_event/modules/oe_content_event_event_programme/src/Entity/ProgrammeItemType.php]:
```suggestion
 *   config_prefix = "oe_event_programme_type",
```

File: modules/oe_content_event/modules/oe_content_event_event_programme/src/Entity/ProgrammeItemType.php [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+ *       "edit" = "Drupal\oe_content\Form\EntityTypeForm",
+ *       "delete" = "Drupal\oe_content\Form\EntityTypeDeleteForm",
+ *     },
+ *     "route_provider" = {
+ *       "html" = "Drupal\Core\Entity\Routing\AdminHtmlRouteProvider",
+ *     },
+ *   },
+ *   admin_permission = "manage corporate content entity types",
+ *   links = {
+ *     "add-form" = "/admin/structure/event_programme_type/add",
```
Comment (yenyasinn) [inline modules/oe_content_event/modules/oe_content_event_event_programme/src/Entity/ProgrammeItemType.php]:
```suggestion
 *     "add-form" = "/admin/structure/oe_event_programme_type/add",
```

File: modules/oe_content_event/modules/oe_content_event_event_programme/src/PermissionCallbacks.php [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+/**
+ * Provides dynamic permissions for the Programme Item entity.
+ */
+class PermissionCallbacks extends PermissionCallbacksBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function getEntityTypeId(): string {
+    return 'event_programme';
```
Comment (yenyasinn) [inline modules/oe_content_event/modules/oe_content_event_event_programme/src/PermissionCallbacks.php]:
```suggestion
    return 'oe_event_programme';
```

File: modules/oe_content_event/modules/oe_content_event_event_programme/src/Entity/ProgrammeItem.php [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+ *   },
+ *   bundle_entity_type = "oe_event_programme_type",
+ *   field_ui_base_route = "entity.oe_event_programme_type.edit_form",
+ *   links = {
+ *     "canonical" = "/admin/content/oe_event_programme/{oe_event_programme}",
+ *     "add-page" = "/admin/content/oe_event_programme/add",
+ *     "add-form" = "/admin/content/oe_event_programme/add/{oe_event_programme_type}",
+ *     "edit-form" = "/admin/content/oe_event_programme/{oe_event_programme}/edit",
+ *     "delete-form" = "/admin/content/oe_event_programme/{oe_event_programme}/delete",
+ *     "collection" = "/admin/content/oe_vent_programme",
```
Comment (yenyasinn) [inline modules/oe_content_event/modules/oe_content_event_event_programme/src/Entity/ProgrammeItem.php]:
```suggestion
 *     "collection" = "/admin/content/oe_event_programme",
```

File: modules/oe_content_event/oe_content_event.post_update.php [hunk 17]
```diff
@@ -96,10 +96,10 @@ function oe_content_event_post_update_00005(): void {
 }
 
 /**
- * Add Media field.
+ * Add new fields.
  */
 function oe_content_event_post_update_20001(): void {
-  $storage = new FileStorage(drupal_get_path('module', 'oe_content_event') . '/config/post_updates/20001_media_field');
+  $storage = new FileStorage(drupal_get_path('module', 'oe_content_event') . '/config/post_updates/20001_event_v2_fields');
```
Comment (yenyasinn) [inline modules/oe_content_event/oe_content_event.post_update.php:103]: Need to enable module `oe_content_event_event_programme` first.

File: modules/oe_content_event/oe_content_event.post_update.php [hunk 18]
```diff
@@ -96,10 +96,10 @@ function oe_content_event_post_update_00005(): void {
 }
 
 /**
- * Add Media field.
+ * Add new fields.
```
Comment (yenyasinn) [inline modules/oe_content_event/oe_content_event.post_update.php]: Might be extend description to point out all new fields: `Add Media and Programme fields`?

Comment (22Alexandra) [reply modules/oe_content_event/oe_content_event.post_update.php]: I think there will be more fields to be added so better to add a comment in the update, ok?

Comment (yenyasinn) [reply modules/oe_content_event/oe_content_event.post_update.php]: We can update comment later also accordingly, if needed. But if there won't be any new fields this comment won't be descriptive.

**Review by yenyasinn — DISMISSED** [review]

File: modules/oe_content_event/modules/oe_content_event_event_programme/config/install/oe_content_event_event_programme.oe_event_programme_type.default.yml [hunk 19]
```diff
@@ -0,0 +1,6 @@
+langcode: en
+status: true
+dependencies: {  }
+id: default
```
Comment (sergepavle) [inline modules/oe_content_event/modules/oe_content_event_event_programme/config/install/oe_content_event_event_programme.oe_event_programme_type.default.yml]: Maybe better to use `oe_default` machine name for consistency with other corporate entity types?

File: modules/oe_content_event/modules/oe_content_event_event_programme/config/install/field.field.oe_event_programme.default.event_programme_dates.yml [hunk 20]
```diff
@@ -0,0 +1,20 @@
+langcode: en
+status: true
+dependencies:
+  config:
+    - field.storage.oe_event_programme.event_programme_dates
+    - oe_content_event_event_programme.oe_event_programme_type.default
+  module:
+    - datetime_range
+id: oe_event_programme.default.event_programme_dates
+field_name: event_programme_dates
```
Comment (sergepavle) [inline modules/oe_content_event/modules/oe_content_event_event_programme/config/install/field.field.oe_event_programme.default.event_programme_dates.yml]: Should we have this field name?
```suggestion
field_name: oe_event_programme_dates
```

File: modules/oe_content_event/modules/oe_content_event_event_programme/src/Entity/ProgrammeItem.php [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+ *   },
+ *   bundle_entity_type = "oe_event_programme_type",
+ *   field_ui_base_route = "entity.oe_event_programme_type.edit_form",
+ *   links = {
+ *     "canonical" = "/admin/content/oe_event_programme/{oe_event_programme}",
+ *     "add-page" = "/admin/content/oe_event_programme/add",
+ *     "add-form" = "/admin/content/oe_event_programme/add/{oe_event_programme_type}",
+ *     "edit-form" = "/admin/content/oe_event_programme/{oe_event_programme}/edit",
+ *     "delete-form" = "/admin/content/oe_event_programme/{oe_event_programme}/delete",
+ *     "collection" = "/admin/content/oe_event_programme",
```
Comment (sergepavle) [inline modules/oe_content_event/modules/oe_content_event_event_programme/src/Entity/ProgrammeItem.php:66]: I think in this sub-module we have missed *.links.task.yml file for possibility to see list of Event Programmes in /admin/content page.

File: tests/src/Behat/Content/Node/EventContentContext.php [hunk 22]
```diff
@@ -67,6 +67,7 @@ public function alterEventFields(BeforeParseEntityFieldsScope $scope): void {
       'Teaser' => 'oe_teaser',
       'Venue' => 'oe_event_venue',
       'Contact' => 'oe_event_contact',
+      'Programme' => 'oe_event_programme',
```
Comment (sergepavle) [inline tests/src/Behat/Content/Node/EventContentContext.php:70]: Looks like also switch part should be updated (should be with 'Venue' and 'Contact' entity types)

File: modules/oe_content_event/modules/oe_content_event_event_programme/src/Entity/ProgrammeItem.php (hunk 21 above)
Comment (22Alexandra) [reply modules/oe_content_event/modules/oe_content_event_event_programme/src/Entity/ProgrammeItem.php:66]: Why would it be needed there? Isn't this supposed to be used only within Event CT?

Comment (sergepavle) [reply modules/oe_content_event/modules/oe_content_event_event_programme/src/Entity/ProgrammeItem.php:66]: Venue entity type as I understand also used only inside Event content type, but the tab is present on /admin/content page.

**Review by sergepavle — DISMISSED** [review]

**Review by sergepavle — APPROVED** [review]

**Review by yenyasinn — APPROVED** [review]

File: modules/oe_content_event/config/install/field.storage.node.oe_event_programme.yml (hunk 1 above)
Comment (22Alexandra) [reply modules/oe_content_event/config/install/field.storage.node.oe_event_programme.yml]: This had to be translatable...

**Pull request merged**